### PR TITLE
audio_renderer: Better voice mixing and 6 channel downmixing

### PR DESCRIPTION
**Pull request by ogniK5377**
Supersedes #3738 and #3321

Expands on the initial work done by tobi to fix issues in other games for the voice mixer. This is still preliminary mixing as we don't have dedicated mix buffers yet. 6 channel downmixing was merged into this PR since it needed to be changed either way for voice mixing.

### CODE REVIEW

File: src/audio_core/audio_renderer.cpp [hunk 1]
```diff
@@ -37,9 +37,11 @@ class AudioRenderer::VoiceState {
     }
 
     void SetWaveIndex(std::size_t index);
-    std::vector<s16> DequeueSamples(std::size_t sample_count, Core::Memory::Memory& memory);
+    std::vector<s16> DequeueSamples(std::size_t sample_count, Core::Memory::Memory& memory,
+                                    std::array<VoiceResourceInformation*, 6> voice_resources);
     void UpdateState();
-    void RefreshBuffer(Core::Memory::Memory& memory);
+    void RefreshBuffer(Core::Memory::Memory& memory,
+                       std::array<VoiceResourceInformation*, 6> voice_resources);
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:44]: 1. 

You may wish to make this a type alias inside the VoiceState class, given it's repeated in several areas. This should also be passed as a const reference.

File: src/audio_core/audio_renderer.cpp [hunk 2]
```diff
@@ -271,7 +280,8 @@ void AudioRenderer::VoiceState::UpdateState() {
     is_in_use = info.is_in_use;
 }
 
-void AudioRenderer::VoiceState::RefreshBuffer(Core::Memory::Memory& memory) {
+void AudioRenderer::VoiceState::RefreshBuffer(
+    Core::Memory::Memory& memory, std::array<VoiceResourceInformation*, 6> voice_resources) {
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:284]: Ditto about passing it as a const reference.

File: src/audio_core/audio_renderer.cpp [hunk 3]
```diff
@@ -37,9 +37,11 @@ class AudioRenderer::VoiceState {
     }
 
     void SetWaveIndex(std::size_t index);
-    std::vector<s16> DequeueSamples(std::size_t sample_count, Core::Memory::Memory& memory);
+    std::vector<s16> DequeueSamples(std::size_t sample_count, Core::Memory::Memory& memory,
+                                    std::array<VoiceResourceInformation*, 6> voice_resources);
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:41]: const reference.

**Comment by legend800** [comment]
In https://github.com/yuzu-emu/yuzu/pull/3321 bgm was broken on Ultra Street Fighter 2. With this pr, bgm is back but very loud out of right speaker, while sfx are fairly quiet. So a few mixing issues remain there. Sounds fine on master.

**Comment by ogniK5377** [comment]
> In #3321 bgm was broken on Ultra Street Fighter 2. With this pr, bgm is back but very loud out of right speaker, while sfx are fairly quiet. So a few mixing issues remain there. Sounds fine on master.

Should be fixed now, I had a little bit of a logic error. It sounds fine with the latest commit

**Comment by legend800** [comment]
Sounds perfect now! Thanks for the quick fix.

**Comment by legend800** [comment]
Sound is Mario Tennis not correct now. Bgm is barely audible with normal sfx volume. No issues prior.

**Comment by FernandoS27** [comment]
Since I don't know if this can be merged. I'm moving it to mainline.

**Comment by ogniK5377** [comment]
Shouldn't be merged yet as it causes regressions in some games(Shovel knight)

**Comment by callmenemo491** [comment]
"yuzu verify build" fails with:
_The pipeline is not valid. Could not get the latest source version for repository yuzu-emu/yuzu hosted on https://github.com/ using ref 818726c1d865be4b14969687302c9d9d4e5b3346. GitHub reported the error, "Bad credentials"_

**Pull request merged**